src/System.Private.CoreLib/src/System/DateTime.Windows.cs

```diff
@@ -0,0 +1,182 @@
+        {
+            FullSystemTime time = new FullSystemTime();
+            if (SystemFileTimeToSystemTime(fileTime, ref time))
+            {
+                time.hundredNanoSecond = fileTime % TicksPerMillisecond;
+                return CreateDateTimeFromSystemTime(ref time);
+            }
+
+            throw new ArgumentOutOfRangeException("fileTime", SR.ArgumentOutOfRange_DateTimeBadTicks);
+        }
+
+        internal static long InternalToFileTime(long ticks)
+        {
+            long fileTime = 0;
+            FullSystemTime time = new FullSystemTime(ticks);
+            if (SystemTimeToSystemFileTime(ref time, ref fileTime))
+            {
+                return fileTime + ticks % TicksPerMillisecond;
+            }
+
+            throw new ArgumentOutOfRangeException(null, SR.ArgumentOutOfRange_FileTimeInvalid);
+        }
+
+        // Just in case for any reason CreateDateTimeFromSystemTime not get inlined,
+        // we are passing time by ref to avoid copying the structure while calling the method.
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        internal static DateTime CreateDateTimeFromSystemTime(ref FullSystemTime time)
+        {
+            long ticks  = DateToTicks(time.wYear, time.wMonth, time.wDay);
+            ticks += TimeToTicks(time.wHour, time.wMinute, time.wSecond);
+            ticks += time.wMillisecond * TicksPerMillisecond;
+            ticks += time.hundredNanoSecond;
+            return new DateTime( ((UInt64)(ticks)) | KindUtc);
+        }
+
+        // FullSystemTime struct matches Windows SYSTEMTIME struct, except we added the extra nanoSeconds field to store
+        // more precise time.
+        [StructLayout(LayoutKind.Sequential)]
+        internal struct FullSystemTime
+        {
+            internal FullSystemTime(int year, int month, DayOfWeek dayOfWeek, int day, int hour, int minute, int second)
+            {
+                wYear = (ushort) year;
+                wMonth = (ushort) month;
+                wDayOfWeek = (ushort) dayOfWeek;
+                wDay = (ushort) day;
+                wHour = (ushort) hour;
+                wMinute = (ushort) minute;
+                wSecond = (ushort) second;
+                wMillisecond = 0;
+                hundredNanoSecond = 0;
+            }
+
+            internal FullSystemTime(long ticks)
+            {
+                DateTime dt = new DateTime(ticks);
+
+                int year, month, day;
+                dt.GetDatePart(out year, out month, out day);
+
+                wYear = (ushort) year;
+                wMonth = (ushort) month;
+                wDayOfWeek = (ushort) dt.DayOfWeek;
+                wDay = (ushort) day;
+                wHour = (ushort) dt.Hour;
+                wMinute = (ushort) dt.Minute;
+                wSecond = (ushort) dt.Second;
+                wMillisecond = (ushort) dt.Millisecond;
+                hundredNanoSecond = 0;
+            }
+
+            internal ushort wYear;
```

Comment:
Would it look better to have actual SYSTEMTIME field here? I looks like it would save us some casts where we pass around `FullSystemTime`, but we are actually passing just `SYSTEMTIME`.

Reply:
will do that.